### DISCUSSION TITLE
Can we hide database, table, schema information in xhr responses?

### DISCUSSION THREAD

**Discussion by m-jooibar** [apache/superset#17396] [Q&A / Help]
Hello Superset team .  
I run superset by docker-compose-non-dev.yml file. 
I saw a strange case in my dashboard.  I created a dashboard and I imported  one chart in it . When I load my dashboard I have several xhr requests in my google chrome network section.  
as you can see in bellow photo we have several xhr request :  

![Screenshot from 2021-11-06 14-10-23](https://user-images.githubusercontent.com/42513136/140608726-eb81356b-a838-4a75-9b1b-21fc40fbe73d.jpg)

i expand two of them for you : 

First :   datasets 
as you can see you give some details that are not necessary for all users. why user can see important details such as table names,columns names and our queries ?  I searched in your docs but I did not find any . 
I guessed maybe if I run superset by docker-compose-non-dev it will run on production mode and we will not see this information but my guess was wrong . 

![Screenshot from 2021-11-06 14-12-22 (1)](https://user-images.githubusercontent.com/42513136/140608736-94fb6d0a-8fc2-4746-98fd-3853e801c13f.jpg)

Second: http://localhost:8088/api/v1/chart/data?form_data=%7B%22slice_id%22%3A1%7D&dashboard_id=1
In this case we have a previous problem too . We are seeing some data that should be hidden.  i know that we can handle this by    result_type  parameter that it can give on of these values [ columns,full,query and ...  ]
but i do not know how we can pass this value to this parameter when we are loading dashboard in Superset or how we can set 'columns' by default for  result_type 

![Screenshot from 2021-11-06 14-29-41](https://user-images.githubusercontent.com/42513136/140608742-69422d8f-16fb-4c9d-9e1c-73d2a02e5d10.jpg)

I hope you answer me .  I am waiting for your answer. Thanks .

**Comment by srinify** [comment]
To add some color from @nytai who responded in Slack, one option could be to "sanitize" the information shown in the frontend based on the permissions the user has.

**Reply by nytai** [reply]
This info is required for showing things like "view SQL query" for chart, etc. I guess the fix here is to check if the permission is there before including the info. This does very much alter the api schema and is a lot of work, but it should be done to enable a truly secure embeddable dashboards use case.

**Reply by m-jooibar** [reply]
> This info is required for showing things like "view SQL query" for chart, etc. I guess the fix here is to check if the permission is there before including the info. This does very much alter the api schema and is a lot of work, but it should be done to enable a truly secure embeddable dashboards use case.


that's true. this information is required but superset can build this query in back without all this details . if this information need to build a query ( for case that we want give result according to filters) , superset should only send filtered information , not all tables and schema and database details !!. whether superset send all information or only send filtered item to back , superset can create full query .all of reporting tools such as metabase and ...  can filter result by user choose , but nobody do not send and show all of information. we have to provide all of condittion for  one safe software.

**Reply by kmzs** [reply]
These detailed API responses pose a security risk. Many details are not necessary and should not be returned especially for embedded dashboards.

[Excessive Data Exposure](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa3-excessive-data-exposure.md) is a common security risk for APIs ranked # 3 on the [OWASP API Security Top 10 2019](https://owasp.org/www-project-api-security/).

We opened a new issue #22960 for this problem (among with another problem) and referenced this discussion.

**Reply by waleedshabbir1** [reply]
is there any update regarding this?

**Comment by junlincc** [comment]
thanks @srinify ! should we add it to our FAQ page?

**Reply by m-jooibar** [reply]
it is not just enough to add to FAQ page . superset have to find a good solution for this issue .

**Reply by srinify** [reply]
yeah this doesn't feel like a FAQ thing :). The bigger question is how we should think about what query information is / isn't visible to the frontend.

**Reply by m-jooibar** [reply]
thank you very much

**Comment by falgunifb** [comment]
Any update on this?

**Reply by Rudneer** [reply]
Any update will be really helpful @junlincc

**Comment by asishPatnaik2000** [comment]
Any update on this?
